File: src/host.zig
/// Host interface and implementations for Evm
///
/// NOTE: This module provides a minimal host interface for testing purposes.
/// The EVM's inner_call method now uses CallParams/CallResult directly and does not
/// go through this host interface - it handles nested calls internally.
const std = @import("std");
const primitives = @import("voltaire");
const Address = primitives.Address.Address;

/// Minimal host interface for external state access (balances, storage, code, nonces)
/// NOTE: This is NOT used for nested calls - EVM.inner_call handles those directly
pub const HostInterface = struct {
    ptr: *anyopaque,
    vtable: *const VTable,

    pub const VTable = struct {
        getBalance: *const fn (ptr: *anyopaque, address: Address) u256,
        setBalance: *const fn (ptr: *anyopaque, address: Address, balance: u256) void,
        getCode: *const fn (ptr: *anyopaque, address: Address) []const u8,
        setCode: *const fn (ptr: *anyopaque, address: Address, code: []const u8) void,
        getStorage: *const fn (ptr: *anyopaque, address: Address, slot: u256) u256,
        setStorage: *const fn (ptr: *anyopaque, address: Address, slot: u256, value: u256) void,
        getNonce: *const fn (ptr: *anyopaque, address: Address) u64,
        setNonce: *const fn (ptr: *anyopaque, address: Address, nonce: u64) void,
        accountExists: ?*const fn (ptr: *anyopaque, address: Address) bool = null,
        accountHasStorage: ?*const fn (ptr: *anyopaque, address: Address) bool = null,
        deleteAccount: ?*const fn (ptr: *anyopaque, address: Address) void = null,
    };

    pub fn getBalance(self: HostInterface, address: Address) u256 {
        return self.vtable.getBalance(self.ptr, address);
    }

    pub fn setBalance(self: HostInterface, address: Address, balance: u256) void {
        self.vtable.setBalance(self.ptr, address, balance);
    }

    pub fn getCode(self: HostInterface, address: Address) []const u8 {
        return self.vtable.getCode(self.ptr, address);
    }

    pub fn setCode(self: HostInterface, address: Address, code: []const u8) void {
        self.vtable.setCode(self.ptr, address, code);
    }

    pub fn getStorage(self: HostInterface, address: Address, slot: u256) u256 {
        return self.vtable.getStorage(self.ptr, address, slot);
    }

    pub fn setStorage(self: HostInterface, address: Address, slot: u256, value: u256) void {
        self.vtable.setStorage(self.ptr, address, slot, value);
    }

    pub fn getNonce(self: HostInterface, address: Address) u64 {
        return self.vtable.getNonce(self.ptr, address);
    }

    pub fn setNonce(self: HostInterface, address: Address, nonce: u64) void {
        self.vtable.setNonce(self.ptr, address, nonce);
    }

    pub fn accountExists(self: HostInterface, address: Address) bool {
        if (self.vtable.accountExists) |exists| {
            return exists(self.ptr, address);
        }
        return self.getBalance(address) != 0 or self.getNonce(address) != 0 or self.getCode(address).len != 0;
    }

    pub fn accountHasStorage(self: HostInterface, address: Address) bool {
        if (self.vtable.accountHasStorage) |has_storage| {
            return has_storage(self.ptr, address);
        }
        return false;
    }

    pub fn deleteAccount(self: HostInterface, address: Address) void {
        if (self.vtable.deleteAccount) |delete| {
            delete(self.ptr, address);
            return;
        }
        self.setBalance(address, 0);
        self.setCode(address, &[_]u8{});
        self.setNonce(address, 0);
    }
};

File: src/host.ts
/**
 * Host interface for EVM external state access
 *
 * NOTE: This module provides a minimal host interface for testing purposes.
 * The EVM's inner_call method now uses CallParams/CallResult directly and does not
 * go through this host interface - it handles nested calls internally.
 *
 * This interface is for external state backends (balances, storage, code, nonces).
 * NOT used for nested calls - EVM.inner_call handles those directly.
 */

/**
 * Ethereum address (20 bytes)
 * Represented as a Uint8Array of length 20
 */
export type Address = Uint8Array;

/**
 * Host interface for external state access
 *
 * Implementations should provide methods to access and modify:
 * - Account balances (u256 values)
 * - Account nonces (u64 values)
 * - Contract code (bytecode)
 * - Storage slots (u256 → u256 mapping)
 */
export interface HostInterface {
  /**
   * Get the balance of an account
   * @param address - Account address
   * @returns Balance as bigint (u256)
   */
  getBalance(address: Address): bigint;

  /**
   * Set the balance of an account
   * @param address - Account address
   * @param balance - New balance as bigint (u256)
   */
  setBalance(address: Address, balance: bigint): void;

  /**
   * Get the code (bytecode) of an account
   * @param address - Account address
   * @returns Contract bytecode as Uint8Array
   */
  getCode(address: Address): Uint8Array;

  /**
   * Set the code (bytecode) of an account
   * @param address - Account address
   * @param code - Contract bytecode as Uint8Array
   */
  setCode(address: Address, code: Uint8Array): void;

  /**
   * Get a storage slot value
   * @param address - Account address
   * @param slot - Storage slot as bigint (u256)
   * @returns Storage value as bigint (u256)
   */
  getStorage(address: Address, slot: bigint): bigint;

  /**
   * Set a storage slot value
   * @param address - Account address
   * @param slot - Storage slot as bigint (u256)
   * @param value - Storage value as bigint (u256)
   */
  setStorage(address: Address, slot: bigint, value: bigint): void;

  /**
   * Get the nonce of an account
   * @param address - Account address
   * @returns Nonce as bigint (u64)
   */
  getNonce(address: Address): bigint;

  /**
   * Set the nonce of an account
   * @param address - Account address
   * @param nonce - New nonce as bigint (u64)
   */
  setNonce(address: Address, nonce: bigint): void;
}

/**
 * Helper to create an Address from a hex string
 * @param hex - Hex string (with or without 0x prefix)
 * @returns Address as Uint8Array
 */
export function addressFromHex(hex: string): Address {
  const cleaned = hex.startsWith('0x') ? hex.slice(2) : hex;
  if (cleaned.length !== 40) {
    throw new Error(`Invalid address length: expected 40 hex chars, got ${cleaned.length}`);
  }
  const bytes = new Uint8Array(20);
  for (let i = 0; i < 20; i++) {
    bytes[i] = parseInt(cleaned.slice(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

/**
 * Helper to convert an Address to a hex string
 * @param address - Address as Uint8Array
 * @returns Hex string with 0x prefix
 */
export function addressToHex(address: Address): string {
  return '0x' + Array.from(address).map(b => b.toString(16).padStart(2, '0')).join('');
}

/**
 * Helper to compare two addresses for equality
 * @param a - First address
 * @param b - Second address
 * @returns true if addresses are equal
 */
export function addressEquals(a: Address, b: Address): boolean {
  if (a.length !== 20 || b.length !== 20) return false;
  for (let i = 0; i < 20; i++) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}
